service/pricing/cid_gravity.go

```diff
@@ -0,0 +1,185 @@
+package pricing
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	pb "github.com/textileio/bidbot/gen/v1"
+	golog "github.com/textileio/go-log/v2"
+)
+
+const (
+	// If loading rules from CID gravity API takes longer than this timeout, turn it into background.
+	cidGravityLoadRulesTimeout = 5 * time.Second
+)
+
+var (
+	log              = golog.Logger("bidbot/pricing")
+	cidGravityAPIUrl = "https://api.cidgravity.com/api/integrations/bidbot"
+	// Use cached rules if they are loaded no earlier than this period.
+	cidGravityCachePeriod = time.Minute
+)
+
+// CIDGravityRules is the format CID gravity API returns. Needs to be public to unmarshal JSON payload. Do not use.
+type CIDGravityRules struct {
+	Blocked         bool
+	MaintenanceMode bool
+	PricingRules    []struct {
+		Verified    bool
+		MinSize     uint64 // bytes
+		MaxSize     uint64 // bytes
+		MinDuration uint64 // epoches
+		MaxDuration uint64 // epoches
+		Price       int64  // attoFil
+	}
+}
+
+type cidGravityRules struct {
+	apiKey      string
+	clientRules map[string]*clientRules
+	lkRules     sync.Mutex
+}
+
+// NewCIDGravityRules returns PricingRules based on CID gravity configuration for the storage provider.
+func NewCIDGravityRules(apiKey string) PricingRules {
+	return &cidGravityRules{apiKey: apiKey, clientRules: make(map[string]*clientRules)}
+}
+
+// PricesFor looks up prices for the auction based on its client address.
+func (cg *cidGravityRules) PricesFor(auction *pb.Auction) (prices ResolvedPrices, valid bool) {
+	cg.lkRules.Lock()
+	rules, exists := cg.clientRules[auction.ClientAddress]
+	if !exists {
+		rules = newClientRulesFor(cg.apiKey, auction.ClientAddress)
+	}
+	cg.clientRules[auction.ClientAddress] = rules
+	cg.lkRules.Unlock()
+	return rules.PricesFor(auction)
+}
+
+type clientRules struct {
+	apiKey           string
+	clientAddress    string
+	rules            atomic.Value // *CIDGravityRules
+	rulesLastUpdated atomic.Value // time.Time
+	rulesETag        atomic.Value // string
```

Comment:
`lkLoadRules` is to sequentialize the call to `maybeReloadRules`, and the atomic values are preventive to prevent goroutine taken longer than 5s from conflicting with the next call to `maybeReloadRules`, which can hardly happen in reality. In this case, `rules` and `rulesLastUpdated` being out of sync is not a big problem, but you have a great point about the cautious use of `atomic.Value`.